src/asm.rs

```diff
@@ -0,0 +1,268 @@
+#![allow(unused_macros)]
+
+//! Assembly support.
+//!
+//! ## Startup Code
+//!
+//! When the `on_gba` feature is active, this module includes the `_start`
+//! function as well as the default `_asm_runtime_irq_handler` that it sets.
+//! Both of which are defined within [`global_asm!`][inline_asm] blocks. They
+//! are not intended to be called directly from Rust, and so foreign function
+//! definitions for them are not exposed to Rust in this module.
+//!
+//! [inline_asm]:
+//!     https://doc.rust-lang.org/nightly/reference/inline-assembly.html
+//!
+//! ## Assembly Helpers
+//!
+//! This module also includes a number of functions to allow you to force the
+//! generation of particular assembly instructions that Rust and/or LLVM does
+//! not otherwise make easy to generate.
+
+use bracer::*;
+
+/// Inserts a `nop` instruction.
+#[inline]
+#[cfg_attr(not(feature = "on_gba"), track_caller)]
+pub fn nop() {
+  on_gba_or_unimplemented! {
+    unsafe {
+      core::arch::asm! {
+        "nop",
+      }
+    }
+  }
+}
+
+/// Atomically swap `x` and the 32-bit value stored at `ptr`.
+///
+/// ## Safety
+/// This both reads and writes `ptr`, so all the usual rules of that apply.
+#[inline]
+#[cfg_attr(feature = "on_gba", instruction_set(arm::a32))]
+pub unsafe fn swp(mut ptr: *mut u32, x: u32) -> u32 {
+  on_gba_or_unimplemented! {
+    let output: u32;
+    // Note(Lokathor): This won't actually alter the pointer register, but we
+    // *tell* LLVM that it will because the pointer register can't be used as
+    // the output of the swapping operation.
```

Comment:
Isn't this what the `late` qualifiers are for? To my knowledge, simply specifying `addr` as simply `in` should prevent the compiler allocating that register for any of the other parameters.